File: tree/226_invert_binary_tree.cc
#include "tree_node.h"

// 2024/01/09: 自己写出来的版本.
TreeNode* invertTree(TreeNode* root) {
  if (root == nullptr) {
    return root;
  }
  TreeNode* temp = root->left;
  root->left = root->right;
  root->right = temp;
  invertTree(root->left);
  invertTree(root->right);
  return root;
}

// 官方答案
TreeNode* invertTree(TreeNode* root) {
  if (root == nullptr) {
    return root;
  }
  TreeNode* left = invertTree(root->left);
  TreeNode* right = invertTree(root->right);
  root->left = right;
  root->right = left;
  return root;
}

File: tree/226_invert_binary_tree.py
from typing import Optional
from tree.tree_node import TreeNode, create_tree_node, print_tree_node


# 2026/05/18: 自己写出来, 根据答案进行简化.
def invert_tree(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if root is None:
        return None

    left = invert_tree(root.left)
    right = invert_tree(root.right)
    root.left = right
    root.right = left
    return root


root = [4, 2, 7, 1, 3, 6, 9]
root = create_tree_node(root)
print_tree_node(root)
root = invert_tree(root)
print_tree_node(root)
